### PR TITLE
[Candidate List] Added Study Consent column

### PR DESCRIPTION
**Pull request by AnyhowStep**
In reference to Redmine ticket 12208,

> it would be helpful (perhaps if it was configurable) if study consent was a column in the candidates page in access profile

No. It will be shown forever more, muahahahahahahahaha *(Until someone tell me "No")*

Added the "Study Consent" column.

If the candidate does not have an associated `participant_status` row, the result for `study_consent` will be `null`. In that case, it defaults to displaying "Not Answered".

### CODE REVIEW

**Comment by AnyhowStep** [comment]
I'm going to work on adding additional columns that show up for each additional type of status. Honest opinion, though, the current consent implementation is... Mortifying.

**Comment by AnyhowStep** [comment]
Additional columns show up, now. However, the UI does not bother formatting the results.

If the database has the value `not_answered`, it will display `not_answered` without performing name-mangling.

The reason is that the current UI code isn't being given enough information to determine if it is processing a "consent" column or not. There are **hacks** but I refuse to perform them on my discretion.

Funnily enough, the reason why the UI code cannot reason about it **reliably** is because the back-end code mangles the column names and doesn't pass the appropriate information.

As such, making the UI for this feature render *nicely* is actually part of a more involved feature.

**Comment by johnsaigle** [comment]
Related: we should make a ticket to discuss hiding columns in React (PR meeting discussion)

File: modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc [hunk 1]
```diff
@@ -77,6 +77,28 @@ class NDB_Menu_Filter_Candidate_List extends NDB_Menu_Filter
...
+        //getConsentStatusFields()
+        $consent         = $config->getSetting('ConsentModule');
+        $consent_details = Utility::asArray($consent['Consent']);
+        if (!$consent_details[0]) {
+            // If only one consent, need to put in an array
+            $temp            = array();
+            $temp[]          = $consent_details;
+            $consent_details = $temp;
+        }
+        foreach ($consent_details as $consentType) {
```
Comment (driusan) [inline modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc]: Shouldn't this check the useConsent config option before adding these columns?

**Comment by AnyhowStep** [comment]
I am reminded of this xkcd:
https://xkcd.com/1171/

**Comment by AnyhowStep** [comment]
`NDB_Menu_Filter` was modified because escaping aliases with backticks is valid MySQL syntax. Also, we really shouldn't be trying to parse user-friendly column names from SQL queries... And especially not with regular expressions...

**Comment by MounaSafiHarab** [comment]
Did basic testing with `useConsent` in config.xml set to true and the column shows up in candidate_list, and setting it to false, the column disappears. Column name is that set in `label` of the config.xml.
So passing manual test.

**Comment by johnsaigle** [comment]
@AnyhowStep The code looks good. Would you mind adding a small comment above the regex describing what it's matching for? It hurts my head to try to figure it out and it would be nice to have a record of your backtick change going forward.

**Comment by AnyhowStep** [comment]
I'll mention it here but I'll also add the comment.

Personal opinion: We really shouldn't be using regex to parse programmer-readable text to user-readable text. This is what the `NDB_Menu_Filter` code is trying to do, which introduces a lot of UI-related problems.

Anyway, the original regex was,
`/[0-9A-Za-z_]+$/`
Which matches strings *ending* with one or more alphanumeric characters and underscores.

It matches strings like,
```
1 AS num, matches num
PSCID AS The_Psc_Id_999, matches The_Psc_Id_999
(SELECT column FROM table LIMIT 1) AS some_value, matches some_value
```

However, that above regex does not take into account escaped column names.
Column names are escaped with backticks. As such, the above can be rewritten,

```
1 AS `num`
PSCID AS `The_Psc_Id_999`
(SELECT column FROM table LIMIT 1) AS `some_value`
```

And this will "break" the regex because none of those strings match *ENDING* in alphanumeric/underscores.

This will cause the column to have an empty string for it's name (or null, actually. But renders as an empty string).

This *HACK* tacks on the parentheses and `` `? ``, meaning that it could optionally end with a trailing backtick and it won't be included in `$match[1]`.

However, this will still break for the following,
`````
1 AS `a column name`, matches name
PSCID AS `A name ending with an escaped backtick```, no match
(SELECT column FROM table LIMIT 1) AS `some``value`, matches value
`````

Honestly, regexes introduce more problems than they solve, in this case =(

**Comment by johnsaigle** [comment]
Thanks for the feedback.

**Comment by ridz1208** [comment]
@driusan changes have been made please re-review

**Comment by driusan** [comment]
Adding the "Blocked" tag because this PR needs someone to adopt it.

(I'll save my comments on what should change until someone picks it up, so that the notifications go to the right person..)

**Comment by johnsaigle** [comment]
@driusan Please assign this to me. Should be simple enough given that this has already been tested and approved.

What changes?

**Comment by johnsaigle** [comment]
@llevitis is taking over the changes instead of me

**Comment by llevitis** [comment]
@driusan the actual 'study consent' column isn't showing up properly now even with useConsent set to true in config.xml ...shall debug

File: modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc [hunk 2]
```diff
@@ -77,6 +77,34 @@ class NDB_Menu_Filter_Candidate_List extends NDB_Menu_Filter
...
+        $consent = $config->getSetting('ConsentModule');
+        //String comparison is required, if ("false") { echo "is true"; }
+        if ($consent['useConsent'] == "true") {
+            $consent_details = $consent['Consent'];
+            //This check is necessary because of the following,
+            //1. Only one consent is configured, an assoc array is returned
+            //2. More than one consent is configured, a numeric array of assoc
+            //   array elements is returned
+            //We want just the numeric array of assoc array elements all the time.
+            if (!isset($consent_details[0])) {
```
Comment (ridz1208) [inline modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc]: use Utility::toArray()

File: modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc [hunk 3]
```diff
@@ -77,6 +77,34 @@ class NDB_Menu_Filter_Candidate_List extends NDB_Menu_Filter
...
+                $name  = Database::singleton()->escape($consentType["name"]);
+                $label = Database::singleton()->escape($consentType["label"]);
+                //This is not foolproof.
+                //This is only done because the following does not work,
+                //"ps.{$name} AS `{$label}`" (notice the backticks `)
+                //The code that creates the tables on the front end will freak out
+                //if it sees a space in the label name
+                $label = preg_replace("/\s/", "_", $label);
+
+                $this->columns[] = "ps.{$name} AS {$label}";
```
Comment (ridz1208) [inline modules/candidate_list/php/NDB_Menu_Filter_candidate_list.class.inc]: causes a problem when there is a dash (-) in the $label.

**Comment by ridz1208** [comment]
@llevitis @driusan CCNA uses multiple consents, like 5 or more this really crowds up the candidate list table, I think they should be concatenated into one column or not displayed at all

**Pull request closed without merging**